File: python/pascals-triangle/pascals_triangle.py
def rows(row_count: int) -> list[list[int]]:
    # Check for invalid input (negative row_count)
    if row_count < 0:
        raise ValueError("number of rows is negative")
    
    # Base case 1: If row_count is 0, return an empty list (empty triangle)
    if row_count == 0:
        return []
    
    # Base case 2: If row_count is 1, return a list containing the first row [1]
    if row_count == 1:
        return [[1]]

    # Recursive case:
    # - Calculate the previous rows using recursion (excluding the last row)
    # - Compute the new row based on the last row
    # - Return the concatenation of previous rows, last row, and the new row
    *prefix, last = rows(row_count - 1)  # Recursive call to get previous rows
    new = [a + b for a, b in zip(last + [0], [0] + last)]  # Compute the new row

    # Return the concatenation of previous rows, last row, and the new row
    return [*prefix, last, new]

File: javascript/pascals-triangle/pascals-triangle.js
//
// This is only a SKELETON file for the 'Pascals Triangle' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const rows = (numRows) => {
  const triangle = []; // Initialize an empty array to represent Pascal's Triangle

  for (let i = 0; i < numRows; i++) {
    triangle[i] = []; // Initialize a sub-array for each row in the triangle

    for (let j = 0; j < i + 1; j++) {
      // Iterate through each element in the current row
      if (j === 0 || j === i) {
        // If it's the first or last element of the row, set it to 1
        triangle[i][j] = 1;
      } else {
        // Otherwise, calculate the value based on the values from the previous row
        triangle[i][j] = triangle[i - 1][j - 1] + triangle[i - 1][j];
      }
    }
  }

  return triangle; // Return the complete Pascal's Triangle
};